File: source/frontend/src/lib/contract.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { api } from './api'
import type { TransactionRead } from './accountHistory'
import type { TransactionCategory } from './transaction'

export const CONTRACT_FREQUENCIES = [
  'WEEKLY',
  'BIWEEKLY',
  'MONTHLY',
  'QUARTERLY',
  'YEARLY',
] as const
export type ContractFrequency = (typeof CONTRACT_FREQUENCIES)[number]

export const CONTRACT_FREQUENCY_FILTERS = [...CONTRACT_FREQUENCIES, 'NONE'] as const
export type ContractFrequencyFilter = (typeof CONTRACT_FREQUENCY_FILTERS)[number]

export type ContractSource = 'DETECTED' | 'MANUAL'
export type ContractAssignment = 'AUTO' | 'MANUAL' | 'EXCLUDED'

export interface ContractRead {
  id: number
  account_id: number
  name: string
  note: string | null
  category: TransactionCategory | null
  source: ContractSource
  median_amount: number | null
  amount_spread: number | null
  frequency: ContractFrequency | null
  interval_days: number | null
  expected_next_date: string | null
  is_overdue: boolean
  member_count: number
  amount_per_day: number | null
  amount_per_frequency: Record<ContractFrequency, number> | null
}

export interface ContractMemberRead extends TransactionRead {
  contract_assignment: ContractAssignment | null
  is_outlier: boolean
}

export interface ContractDetailRead extends ContractRead {
  members: ContractMemberRead[]
}

export interface ContractFilters {
  account_ids?: number[]
  amount_from?: number
  amount_to?: number
  categories?: TransactionCategory[]
  frequencies?: ContractFrequencyFilter[]
  overdue?: boolean
}

export function filterContracts(
  contracts: ContractRead[],
  filters: ContractFilters,
): ContractRead[] {
  const { account_ids, amount_from, amount_to, categories, frequencies, overdue } = filters
  // A facet is inactive only when its key is absent. A present-but-empty array
  // means "none selected" and matches nothing (the "Keine" button).
  return contracts.filter((contract) => {
    if (account_ids && !account_ids.includes(contract.account_id)) return false
    if (categories && !(contract.category && categories.includes(contract.category))) return false
    if (frequencies && !frequencies.includes(contract.frequency ?? 'NONE')) return false
    if (amount_from !== undefined && (contract.median_amount ?? -Infinity) < amount_from)
      return false
    if (amount_to !== undefined && (contract.median_amount ?? Infinity) > amount_to) return false
    if (overdue && !contract.is_overdue) return false
    return true
  })
}

export type ContractCostPeriod = 'DAY' | 'WEEKLY' | 'MONTHLY' | 'YEARLY'
export const CONTRACT_COST_PERIODS: ContractCostPeriod[] = ['DAY', 'WEEKLY', 'MONTHLY', 'YEARLY']

export function contractAmountForPeriod(
  contract: ContractRead,
  period: ContractCostPeriod,
): number | null {
  if (period === 'DAY') return contract.amount_per_day
  return contract.amount_per_frequency?.[period] ?? null
}

export function sumContractsForPeriod(
  contracts: ContractRead[],
  period: ContractCostPeriod,
): number {
  return contracts.reduce((sum, contract) => {
    const value = contractAmountForPeriod(contract, period)
    return value === null ? sum : sum + value
  }, 0)
}

export const OVERDUE_BANNER_MONTHS = 2

export function monthsOverdue(expectedNextDate: string, now: Date = new Date()): number {
  const due = new Date(`${expectedNextDate}T00:00:00`)
  if (Number.isNaN(due.getTime())) return 0
  let months = (now.getFullYear() - due.getFullYear()) * 12 + (now.getMonth() - due.getMonth())
  if (now.getDate() < due.getDate()) months -= 1
  return Math.max(0, months)
}

export function hasActiveContractFilters(filters: ContractFilters): boolean {
  return Boolean(
    filters.account_ids !== undefined ||
    filters.categories !== undefined ||
    filters.frequencies !== undefined ||
    filters.amount_from !== undefined ||
    filters.amount_to !== undefined ||
    filters.overdue,
  )
}

export interface ContractCreatePayload {
  name: string
  account_id: number
  category?: TransactionCategory | null
  frequency?: ContractFrequency | null
}

export interface ContractUpdatePayload {
  name: string
  category?: TransactionCategory | null
  note?: string | null
  frequency?: ContractFrequency | null
}

export const contractQueryKeys = {
  list: ['contracts'] as const,
  detail: (contractId: number) => ['contracts', contractId] as const,
}

export function useContracts() {
  return useQuery({
    queryKey: contractQueryKeys.list,
    queryFn: () => api<ContractRead[]>('/contracts'),
  })
}

export function useContract(contractId: number) {
  return useQuery({
    queryKey: contractQueryKeys.detail(contractId),
    queryFn: () => api<ContractDetailRead>(`/contracts/${contractId}`),
  })
}

export function useCreateContract() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (payload: ContractCreatePayload) =>
      api<ContractDetailRead>('/contracts', { method: 'POST', body: payload }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contractQueryKeys.list })
    },
  })
}

export function useUpdateContract(contractId: number) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (payload: ContractUpdatePayload) =>
      api<ContractDetailRead>(`/contracts/${contractId}`, { method: 'PATCH', body: payload }),
    onSuccess: (updated) => {
      queryClient.setQueryData(contractQueryKeys.detail(contractId), updated)
      queryClient.setQueryData<ContractRead[]>(contractQueryKeys.list, (old) =>
        old
          ? old.map((contract) =>
              contract.id === updated.id
                ? {
                    ...contract,
                    name: updated.name,
                    category: updated.category,
                    frequency: updated.frequency,
                    interval_days: updated.interval_days,
                    expected_next_date: updated.expected_next_date,
                    is_overdue: updated.is_overdue,
                    amount_per_day: updated.amount_per_day,
                    amount_per_frequency: updated.amount_per_frequency,
                  }
                : contract,
            )
          : old,
      )
      queryClient.invalidateQueries({ queryKey: ['account'] })
    },
  })
}

export function useDeleteContract() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (contractId: number) => api<void>(`/contracts/${contractId}`, { method: 'DELETE' }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contractQueryKeys.list })
      queryClient.invalidateQueries({ queryKey: ['account'] })
    },
  })
}

export interface SetTransactionContractArgs {
  transactionId: number
  fromContractId: number | null
  toContractId: number | null
}

export function useSetTransactionContract() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ transactionId, fromContractId, toContractId }: SetTransactionContractArgs) =>
      toContractId !== null
        ? api<ContractDetailRead>(`/contracts/${toContractId}/transactions`, {
            method: 'POST',
            body: { transaction_id: transactionId },
          })
        : api<ContractDetailRead>(`/contracts/${fromContractId}/transactions/${transactionId}`, {
            method: 'DELETE',
          }),
    onSuccess: (_updated, { fromContractId, toContractId }) => {
      queryClient.invalidateQueries({ queryKey: contractQueryKeys.list })
      queryClient.invalidateQueries({ queryKey: ['account'] })
      if (fromContractId !== null)
        queryClient.invalidateQueries({ queryKey: contractQueryKeys.detail(fromContractId) })
      if (toContractId !== null)
        queryClient.invalidateQueries({ queryKey: contractQueryKeys.detail(toContractId) })
    },
  })
}

File: source/backend/api/schemas/contracts/contract.py
import datetime

from pydantic import BaseModel, ConfigDict

from source.backend.api.schemas.transactions.transaction import TransactionRead
from source.backend.helpers import utc_now
from source.backend.models.contracts.contract import Contract
from source.backend.models.contracts.contract_assignment import ContractAssignment
from source.backend.models.contracts.contract_frequency import ContractFrequency
from source.backend.models.contracts.contract_source import ContractSource
from source.backend.models.transactions.transaction_category import TransactionCategory


class ContractCreate(BaseModel):
    name: str
    account_id: int
    category: TransactionCategory | None = None
    frequency: ContractFrequency | None = None


class ContractUpdate(BaseModel):
    name: str
    category: TransactionCategory | None = None
    note: str | None = None
    frequency: ContractFrequency | None = None


class ContractMemberRead(TransactionRead):
    contract_assignment: ContractAssignment | None
    is_outlier: bool


class ContractRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    account_id: int
    name: str
    note: str | None = None
    category: TransactionCategory | None
    source: ContractSource
    median_amount: float | None
    amount_spread: float | None
    frequency: ContractFrequency | None
    interval_days: int | None
    expected_next_date: datetime.date | None
    is_overdue: bool = False
    member_count: int = 0
    amount_per_day: float | None = None
    amount_per_frequency: dict[ContractFrequency, float] | None = None

    @classmethod
    def from_contract(cls: type["ContractRead"], contract: Contract) -> "ContractRead":
        instance = cls.model_validate(contract)
        instance.is_overdue = contract.is_overdue_on(today=utc_now().date())
        instance.member_count = len(contract.members())

        anchor_days = contract.frequency.interval_days if contract.frequency else contract.interval_days
        if contract.median_amount is not None and anchor_days:
            amount_per_day = contract.median_amount / anchor_days
            instance.amount_per_day = amount_per_day
            instance.amount_per_frequency = {
                frequency: amount_per_day * frequency.interval_days for frequency in ContractFrequency
            }
            if contract.frequency:
                instance.amount_per_frequency[contract.frequency] = contract.median_amount

        return instance


class ContractDetailRead(ContractRead):
    members: list[ContractMemberRead] = []

    @classmethod
    def from_contract(cls: type["ContractDetailRead"], contract: Contract) -> "ContractDetailRead":
        members = sorted(contract.members(), key=lambda transaction: transaction.date, reverse=True)
        return cls(
            **ContractRead.from_contract(contract).model_dump(),
            members=[
                ContractMemberRead(
                    **TransactionRead.model_validate(transaction).model_dump(),
                    contract_assignment=transaction.contract_assignment,
                    is_outlier=contract.is_outlier(transaction),
                )
                for transaction in members
            ],
        )


class ContractAssignRequest(BaseModel):
    transaction_id: int
